### PR TITLE
rsx: MGS4 fixes

### PR DESCRIPTION
**Pull request by kd-11**
- Implement 24-bit floats in depth texture decode step [D24S8->ARGB8]
- Fix some shader decompiler bugs
- Fix behavior of varying registers. Not all registers are created equal. Some have in-built saturate modifiers and texN registers can have their behavior altered through other registers. All tests confirmed on real hardware.

### CODE REVIEW

**Comment by AniLeo** [comment]
Fixes Lord of the Rings: Conquest, excellent :+1:

**Comment by kd-11** [comment]
NOTE: This branch requires https://github.com/RPCS3/rpcs3/pull/6376 to be merged first to resolve some unrelated bugs in master. Without the fixes rpcs3 may crash or have unexpected graphics glitches.

**Comment by Xcedf** [comment]
Not only MGS4, MGSV Ground Zeroes showing graphics now

**Comment by digitaldude555** [comment]
Screenshot (491)
Infamous glide graphics changed.

**Comment by Whatcookie** [comment]
Fixes some issues in ninja gaiden games:

Fixed waterfall being drawn ontop of other stuff
<details>
<summary>NGS2</summary>

Master
screenshot-2019_08_21_14_10_33

PR
screenshot-2019_08_21_13_52_24

</details>


Fixed depth of field
<details>
<summary>NG3:RE</summary>

Master
screenshot-2019_08_21_14_06_06

PR
screenshot-2019_08_21_13_41_57

</details>

**Comment by Asinin3** [comment]
There are currently some major issues in a lot of games including Persona 5, I'd advise testers to wait for new commits before testing games other than MGS4.

**Comment by ju2ef1986** [comment]
MGS 4 is crashing when you select difficulty level. This is with SPU cache enabled. With cache disabled you can get all the way to the "war has changed" bit but there is no sound and it crashes right away. Let me know if you need any logs etc. Or if you want me to try different settings 

i7 5820 GTX 1070 16GB RAM SSD Windows 10 1903

**Comment by kd-11** [comment]
Graphics problems in the other games should be fixed now.

**Comment by Snakegodeater** [comment]
> MGS 4 is crashing when you select difficulty level. This is with SPU cache enabled. With cache disabled you can get all the way to the "war has changed" bit but there is no sound and it crashes right away. Let me know if you need any logs etc. Or if you want me to try different settings
> 
> i7 5820 GTX 1070 16GB RAM SSD Windows 10 1903

Yah the game had issues at for me too With SPU LLVM and With SPU ASMJIT too. Even with the previews weeks builds . On Asmij you get no sound In-game but u can play A little before it Freezes The emulator. With SPU LLVM U get crashes when u enter the game. For a few minutes but with sound . Then when u restart the emulator and the game it just Freezes on Difficulty selection !

**Comment by ju2ef1986** [comment]
Looks like it fixed one issue from GT6 The text now appear with reflection instead of the glitched look
PR
image
MASTER
image

**Comment by Snakegodeater** [comment]
On MGS4 Demo i get This Error 
F {RSX [0x0d0575c]} RSX: class std::runtime_error thrown: Unknown index array type 116
(in file c:\projects\rpcs3\rpcs3\emu\rsx\gcm_enums.cpp:27)
On latest comit  
4790k+R9 380 Graphics although seems fixed

**Comment by incognitoh** [comment]
slight regression:  The Bigs 2

Master:
bigsmster

PR:
bigsmgs4

basically some square shadows appear that shouldnt be there..and I did clear the cache before testing.

gpu: gtx1070

**Comment by incognitoh** [comment]
Bigs 2 (and a couple others) now fixed with RPCS3 v0.0.6-7bd329e0.  thanks

**Comment by MarioSonic2987** [comment]
Fixes excessive blur in Just Cause 2.

**Before**:
image

**After**:
image

**Comment by jenci8888** [comment]
Blur game "blur" has been fixed on the latest commit! (Needs WCB, RCB and WDB)
Plus it ran full speed with low framerate compared to slower speed on master!
image

**Comment by kd-11** [comment]
Did a partial revert on the fix. This PR has been relegated to WIP pending more hardware test results.

**Comment by dyno0** [comment]
This PR affects The Evil Within white screen without forcing CPU blit emulation (still needs WCB). On master build the white screen is consistent, where on this PR it occasionally renders the whole image for a split second, in menu and ingame.

**Comment by stride21** [comment]
Before:
before battlefield
After:
bf pr fire

**Comment by arcadee1977** [comment]
Also fixed NGS2 water surface effect 👍 
BEFORE:

未标题-3

AFTER:
未标题-1

BEFORE:
未标题-4

AFTER:
未标题-2

**Comment by Xcedf** [comment]
GoW 3 Regression; previously was in this PR than it was fixed by 87bb9cb, now it regressed back after f6322cb
Before:
screenshot-2019_08_21_23_27_31
After:
screenshot-2019_08_21_23_24_26

**Comment by Asinin3** [comment]
kd had to re-break some of the fixes he made to bring back fixed shadows in MGS4. He said he will look at it some more tomorrow. Currently Ni No Kuni and Persona 5 have regressed back to the way it was in earlier commits as well. But _most_ of the regressions are fixed.

**Comment by kd-11** [comment]
@Xcedf Retry with the newest commit.
After running some more tests, it appears the extra texture register 9 behaves differently depending on how it is used. Attempting to load its contents into other registers does not work (always returns vector4(0, 0, 0, 1)) but if used with a sampling instruction the texture unit is able to access it. Likely a hardware bug or a provisioning to zero initialize a register, but likely just a hw bug lol.

**Comment by Xcedf** [comment]
@kd-11, not fixed unfortunately.

**Comment by kd-11** [comment]
Ok, I think I know what's wrong.

**Comment by Kennedy88** [comment]
Fixes Siren blood curse

Before
CRrMHgt

After
k10kA2r

**Comment by legend800** [comment]
Looks like this PR affects this issue:  https://github.com/RPCS3/rpcs3/issues/6304 The headlights are broken on Master but outright missing on this PR. 

Master:
master
master3

PR:
pr1
pr2

Real:
rea

**Comment by kd-11** [comment]
Pushed commit to try and match hardware more closely. Breaks shadows in mgs4 again, but all the other regressions should be fixed. If this proves to not break any games a workaround for mgs4 will be made available.

**Comment by Xcedf** [comment]
@kd-11 GoW3 refixed, and MGS4 have black shadows again as you said, but in Killzone 2 main menu now balck screen with sound and subtitles are blacked out, same issue in TLoU and Uncharted 3
screenshot-2019_08_23_21_06_02
screenshot-2019_08_23_21_12_18
screenshot-2019_08_23_21_12_31

**Comment by kd-11** [comment]
Minor bug with some registers. I'm trying to get it sorted out.

**Comment by raveskirza** [comment]
TTT2 text is no longer visible

PR
image

Master
image

**Comment by Snakegodeater** [comment]
Nothing ! MGS4 on Latest PR Still Crashes on SPU LLVM On my R9 380 .Also no audio and when i try to load a level the Display driver crashes when compiling Shaders when the game tries to install the chapter

**Comment by incognitoh** [comment]
The entire MLB the Show series now have a completely black screen for the intros and menus. Here is a pic of the main menu of MLB 16 but the entire series of The Show seems to be affected.
MLB

But The Bigs 2 bug mentioned earlier is fixed again.

**Comment by incognitoh** [comment]
MLB the Shows are now fixed

**Comment by Xcedf** [comment]
Regressions are fixed.

**Comment by Snakegodeater** [comment]
Can u test with AMD Graphics Hw ??still getting the same issues with MGS4. Is it poor AMD drivers or another problem like SPU LLVM OR ASMJIT Recompiler ?

**Comment by Asinin3** [comment]
@Snakegodeater All your posts here are unrelated to this Pull Request and belong on our Discord's #help channel. Issues with the game not booting and such have nothing to do with the graphical improvements kd-11 is making here. https://discord.me/rpcs3

**Comment by stride21** [comment]
Some of the textures in Motorstorm are no longer white. 
Before:
before
After:
after

File: rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp [hunk 1]
```diff
@@ -520,27 +521,81 @@ template<typename T> std::string FragmentProgramDecompiler::GetSRC(T src)
 			"ssa"
 		};
 
-		//TODO: Investigate effect of input modifier on this type
+		// NOTE: Hw testing showed the following:
+		// 1. Reading from registers 1 and 2 (COL0 and COL1) is clamped to (0, 1)
+		// 2. Reading from registers 4-12 (inclusive) is not clamped, but..
+		// 3. If the texcoord control mask is enabled, the last 2 values are always 0 and 1!
+		// 4. Reading from tc9 returns one value in w (point size?) and contains a 1
```
Comment (vesuvian) [inline rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp]: Could this be perspective shear rather than point size?

**Comment by legend800** [comment]
**Motorstorm Apocalypse**: Headlights still no longer render on PR (render but broken on Master). I'm personally fine with it (prefer none to broken), but just any FYI.

**Motorstorm 1**: Appears broken. I get a fatal error (3/3 times) when starting a race.



[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/3536779/RPCS3.zip)

Note: No change to the existing "strips issue" on the track texture:
2019-08-23 19_28_52-Window

**Comment by stride21** [comment]
I tested the same race 10 times clearing the shader cache each time and only had it crash once. I'll add a log just to show settings and PC specs.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3536810/RPCS3.log.gz)

File: rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp (hunk 1 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/Common/FragmentProgramDecompiler.cpp]: Comment is outdated, turns out I had the enable bit for the register in the wrong place. Properly initializing it shows that it does indeed carry information like the other texcoord registers.

**Comment by spyropt** [comment]
i am getting out of memory in mgs4 and garbled image
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3537028/RPCS3.log.gz)
image

**Comment by 13xforever** [comment]
You have gtx 760 with only 2 gb of vram

**Comment by spyropt** [comment]
> You have gtx 760 with only 2 gb of vram

you are telling me that i need more than 2gb vram buffer? or does it have a leak somewher? since it goes from 600/700mb to 2gb in a second

**Comment by kd-11** [comment]
Its a spike that happens when the engine is reconfiguring itself for ingame. Basically throws the surface cache out of balance for a moment but then the garbage collection kicks in and does recover most of the VRAM wasted.

**Comment by kd-11** [comment]
@stride21 @legend800 Any special settings used? It motorstorm runs fine on my end with no errors from the validation layers. If you can make it crash, do so with SDK debug output enabled so I can try and fix it.

**Comment by xYaroslavGTx** [comment]
Fixed menu displaying.
image

**Comment by stride21** [comment]
Motorstorm is consistently crashing while compiling shaders on start up when testing the recent commit.
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/3537348/RPCS3.zip)

**Comment by Xcedf** [comment]
On latest version of PR Heavenly Sword Demo instantly crashes just before ingame on compileing shaders, emu closes itself with no error message no segfault.
Found a little difference in MGS4: on older version of this PR with working shadows ihave this,
screenshot-2019_08_24_19_15_27
in motion the edges of poligons become visible
Current PR shadows black but this scene looks fine
screenshot-2019_08_24_19_23_08

**Comment by legend800** [comment]
@stride21 @kd-11 I just updated to latest build, played Motorstorm (full race), no crashes. Cleared shader cache, played again, no crashes. Same settings as before. Not sure if the changes fixed it or it's just a timing/race condition thing. 

[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3537393/RPCS3.log.gz)

**Comment by Emulator-Team-2** [comment]
@kd-11 
Most games crashes on the latest build  while shaders is compiling.

Naruto Shippuden: Ultimate Ninja Storm Generations

```
[Vulkan]

E {RSX [0x0000000]} RSX: ERROR: 0:161: '=' :  cannot convert from ' const 4-component vector of float' to ' global 4-component vector of float16_t'
ERROR: 0:161: '' : compilation terminated 
ERROR: 2 compilation errors.  No code generated.

E {RSX [0x0000000]} RSX: 
F {RSX [0x0000000]} RSX: class std::runtime_error thrown: Failed to compile fragment shader
(in file C:\projects\rpcs3\rpcs3\Emu\RSX\VK\VKHelpers.h:3153)

```
Log: [RPCS3-VK.zip](https://github.com/RPCS3/rpcs3/files/3537406/RPCS3-VK.zip)

```
[OpenGL]

E {RSX Decompiler Thread} RSX: Failed to compile shader: 0(219) : error C1038: declaration of "wpos" conflicts with previous declaration at 0(218)

```
Log: [RPCS3-GL.zip](https://github.com/RPCS3/rpcs3/files/3537407/RPCS3-GL.zip)

**Comment by kd-11** [comment]
Crashes for GPUs with native fp16 has been fixed. Also added some new findings on register behavior, but I don't expect they will change much.

**Comment by Emulator-Team-2** [comment]
Naruto, inFamous Demo and MGS4-Demo in Vulkan are fixed now in the latest commit.

But OpenGL still crashes with with the same error in MGS4-Demo:
`F {RSX [0x0000000]} RSX: class std::runtime_error thrown: Unknown register name: fogc
(in file C:\projects\rpcs3\rpcs3\Emu\RSX\GL\GLCommonDecompiler.cpp:35)`

Log:[RPCS3-GL.zip](https://github.com/RPCS3/rpcs3/files/3537498/RPCS3-GL.zip)

**Comment by Snakegodeater** [comment]
https://github.com/RPCS3/rpcs3/issues/6403 

I am refrcencing This for AMD  Users !!!

**Comment by hch12907** [comment]
@Snakegodeater You are just annoying and unnecessarily pinging everyone else subscribed to this PR. Please, don't do that.

**Comment by Emulator-Team-2** [comment]
@kd-11 
Dragon Ball: Raging Blast 2 have missing graphics effects on this pr.

PR:

**Link:** [RSX Capture-PR](https://drive.google.com/file/d/1qJUi-cLliduXprWAFKgLcoOKf3o_AhUY/view?usp=sharing)

Master:

**Link:** [RSX Capture-Master](https://drive.google.com/file/d/1IFaTO1E2YULO7SmX57wLNCv5s-PdKnBB/view?usp=sharing)

**Comment by stride21** [comment]
Resistance 3 is crashing with Vulkan on the main menu and when going ingame.
E {RSX [0x39f4e70]} RSX: ERROR: 0:242: 'extraneous semicolon' : not supported for this version or the enabled extensions 
ERROR: 1 compilation errors.  No code generated.


E {RSX [0x39f4e70]} RSX: 
F {RSX [0x39f4e70]} RSX: class std::runtime_error thrown: Failed to compile vertex shader
(in file C:\projects\rpcs3\rpcs3\Emu\RSX\VK\VKHelpers.h:3153)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3537629/RPCS3.log.gz)

**Comment by kamer1337** [comment]
Tony Hawk's Project 8, Tony Hawk's Proving Grounds 
are now spamming
>  RSX: Depth texture bound to pipeline with unexpected format 0x95

**Comment by Xcedf** [comment]
Confirm same spamming reappeared in Uncharted 2
RSX: Depth texture bound to pipeline with unexpected format 0x9C
started with first commit of this PR 3cadd4ce

**Comment by kd-11** [comment]
The spam is correct. The underlying cause was never truly fixed and hiding bugs is not a good approach.

**Comment by kd-11** [comment]
@stride21 You need to provide a shaderlog. Enable "log shader programs" and run until it crashes then attach shaderlog folder along with the logfile. Remember to clear your shaderlog folder beforehand.

**Comment by kd-11** [comment]
@Xcedf @kamer1337 Can I get an RSX capture of affected scenes? Just one will do.

**Comment by stride21** [comment]
@kd-11 Resistance 3 usually gets a trap on its first attempt when trying to go ingame even on master. Should I keep the shaderlog and log info from the first attempt or clear it for when it crashes on the 2nd attempt(when it normally goes ingame on master)?

**Comment by kd-11** [comment]
You don't need to clear shader cache, just shaderlog.

**Comment by stride21** [comment]
I added logs for both the ingame crash on the 2nd attempt and the menu crash. 
[RPCS3 ingame.log.gz](https://github.com/RPCS3/rpcs3/files/3538034/RPCS3.ingame.log.gz)
[shaderlog ingame.zip](https://github.com/RPCS3/rpcs3/files/3538035/shaderlog.ingame.zip)
[RPCS3log menu.zip](https://github.com/RPCS3/rpcs3/files/3538036/RPCS3log.menu.zip)
[shaderlog menu.zip](https://github.com/RPCS3/rpcs3/files/3538037/shaderlog.menu.zip)

**Comment by kd-11** [comment]
Should be fixed now. It was caused by some borked shader that is disabled.

**Comment by Xcedf** [comment]
@kd-11 RSX Capture of U2?

**Comment by stride21** [comment]
Resistance 3 is fixed.

**Comment by AniLeo** [comment]
Shadow intensity on LOTR Conquest now also fixed

screenshot-2019_08_25_21_55_17
screenshot-2019_08_25_21_55_23

**Comment by kd-11** [comment]
@Xcedf Any game with the texture format spam will do.

**Comment by stride21** [comment]
@kd-11 Heres a RSX capture of Uncharted 2 spamming
RSX: Depth texture bound to pipeline with unexpected format 0x9C
[RSX capture](https://www.mediafire.com/file/x3vc4l2erb4e56m/RSX_capture.zip/file)

**Comment by HHCHunter** [comment]
screenshot-2019_08_26_12_46_55

Unsure at the moment if this is a game specific LOD issue or if it's a genuine texture format rendering issue that's currently unimplemented with this PR.

**Comment by SN4KE412** [comment]
Any tutorial on how to use it on MGS4 can't find a way to install it to RPCS3.

**Comment by RainbowCookie32** [comment]
If you need help setting up the emulator, head to our Discord server, don't spam PRs https://discord.me/RPCS3

**Comment by Xcedf** [comment]
@kd-11 U2 capture, before the menu when spamming starts
[NPEA00365_20190826092755_capture.gz](https://github.com/RPCS3/rpcs3/files/3539688/NPEA00365_20190826092755_capture.gz)

**Comment by kd-11** [comment]
I think this leaves only the black shadows in mgs4 as the last bug.

**Comment by Xcedf** [comment]
Yep, no more spamming)

**Comment by Maklai884** [comment]
on 2d0ec734 version in MGS4 - after loading the level, after a few seconds the sound disappears, then after a couple of moments the game freezes. Both with and without sound in the settings.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/3541930/RPCS3.log.gz)

**Comment by kd-11** [comment]
This work is not going to fix runtime issues with the game, its only meant to improve visual output. Note the 'rsx' tag in the heading.

**Comment by kd-11** [comment]
Excessive shadow intensity will be fixed at a later time. It turned out to be complicated to fix it and not break other games, so I'm going to leave if for later.

**Comment by MarioSonic2987** [comment]
Fixes broken models in the following games:
<details>
<summary>Castlevania: Lords of Shadow - Mirror of Fate HD</summary>

image
image
</details>
<details>
<summary>Worms Ultimate Mayhem (long-standing issue since 2016)</summary>

image
image
</details>

**Pull request merged**